File: euclidean_gcd/euclidean_gcd.py
"""
There are two implementations:
Recursive (euclidean_gcd_recursive) and
Non-Recursive (euclidean_gcd)
"""


def euclidean_gcd_recursive(first, second):
    """
    Calculates GCD of two numbers using Euclidean Recursive Algorithm
    :param first: First number
    :param second: Second number
    :return: GCD of the numbers
    """
    if second == 0:
        return first        # First becomes gcd if second becomes zero
    else:
        return euclidean_gcd(second, first % second)


def euclidean_gcd(first, second):
    """
    Calculates GCD of two numbers using Euclidean Iterative Algorithm
    :param first: First number
    :param second: Second number
    :return: GCD of the numbers
    """
    while second != 0:      # Iterate till second becomes zero
        temp = second       # Temporary variable to hold value of second
        second = first % second
        first = temp    # Temporary variable assigned to first
    return first        # When second becomes 0, first becomes gcd of both


def main():
    first = 36
    second = 6
    answer_iterative = euclidean_gcd(first, second)
    answer_recursive = euclidean_gcd_recursive(first, second)
    print("GCD of %d and %d is : %d by recursive algo." % (first, second,
                                                           answer_recursive))
    print("GCD of %d and %d is : %d by iterative algo." % (first, second,
                                                           answer_iterative))


if __name__ == '__main__':
    main()

File: euclidean_gcd/EuclideanGCD.java
/*
 * first --> First number
 * second --> Second number
 * There are two implementations:
 * Recursive(euclideanGCDRecursive) and Non-Recursive(euclideanGCD)
 */

public class EuclideanGCD {
    static int euclideanGCD(int first, int second) {
        while(second != 0) {        // Iterate till second becomes zero
            int temp = second;      // Temporary variable to hold value of second
            second = first % second;
            first = temp;
        }
        return first;       // When second becomes 0, first becomes gcd of both
    }

    static int euclideanGCDRecursive(int first, int second) {
        return (second == 0) ? first : euclideanGCDRecursive(second, (first % second));
                // First becomes GCD when second becomes zero
    }

    public static void main(String[] args) {
        int first = 25;
        int second = 5;
        int answerIterative = EuclideanGCD.euclideanGCD(first, second);
        int answerRecursive = EuclideanGCD.euclideanGCDRecursive(first, second);
        System.out.printf("GCD of %d and %d is : %d by recursive algo.\n", first,
                                                                        second, answerRecursive);
        System.out.printf("GCD of %d and %d is : %d by iterative algo.\n", first,
                                                                        second, answerIterative);
    }
}
